infra/gravity/cluster_status.go

```diff
@@ -2,45 +2,106 @@ package gravity
+func checkNotDegraded(s GravityStatus) error {
+	if s.Cluster.State == constants.ClusterStateDegraded {
+		return trace.CompareFailed("cluster state %q", s.Cluster.State)
 	}
+	if s.Cluster.SystemStatus != constants.SystemStatus_Running {
+		return trace.CompareFailed("expected system_status %v, found %v", constants.SystemStatus_Running, s.Cluster.SystemStatus)
+	}
+	return nil
+}
 
-	err := retry.Do(ctx, func() error {
-		errs := make(chan error, len(nodes))
+// checkActive returns an error if the cluster is degraded or state != active.
+func checkActive(s GravityStatus) error {
+	if err := checkNotDegraded(s); err != nil {
+		return trace.Wrap(err)
+	}
+	if s.Cluster.State != constants.ClusterStateActive {
+		return trace.CompareFailed("expected state %q, found %q", constants.ClusterStateActive, s.Cluster.State)
+	}
+	return nil
+}
 
-		for _, node := range nodes {
-			go func(n Gravity) {
-				_, err := n.Status(ctx)
-				errs <- err
-			}(node)
-		}
+// WaitForActiveStatus blocks until all nodes report state = Active and notDegraded or an internal timeout expires.
+func (c *TestContext) WaitForActiveStatus(nodes []Gravity) error {
+	c.Logger().WithField("nodes", Nodes(nodes)).Info("Waiting for active status.")
+	return c.WaitForStatus(nodes, checkActive)
+}
+
+// WaitForStatus blocks until all nodes satisfy the expected statusValidator or an internal timeout expires.
+func (c *TestContext) WaitForStatus(nodes []Gravity, expected statusValidator) error {
+	b := backoff.NewExponentialBackOff()
+	b.MaxElapsedTime = c.timeouts.ClusterStatus
 
-		err := utils.CollectErrors(ctx, errs)
-		if err == nil {
-			return nil
+	expectStatus := func() (err error) {
+		statuses, err := c.Status(nodes)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		for _, status := range statuses {
+			err = expected(status)
+			if err != nil {
+				c.Logger().WithError(err).WithField("status", status).Warn("Unexpected Status.")
+				return trace.Wrap(err)
+			}
 		}
-		c.Logger().Warnf("Status not available on some nodes, will retry: %v.", err)
-		return wait.Continue("status not ready on some nodes")
-	})
+		return nil
+	}
+
+	err := wait.RetryWithInterval(c.ctx, b, expectStatus, c.Logger())
 
 	return trace.Wrap(err)
+
+}
+
+// Status queries `gravity status` on each node in nodes.
+func (c *TestContext) Status(nodes []Gravity) ([]GravityStatus, error) {
+	ctx, cancel := context.WithTimeout(c.ctx, c.timeouts.NodeStatus)
+	defer cancel()
+
+	errC := make(chan error, len(nodes))
+	valueC := make(chan interface{}, len(nodes))
+
+	for _, node := range nodes {
+		go func(n Gravity) {
+			status, err := n.Status(ctx)
+			errC <- err
+			if status != nil {
+				valueC <- *status
+			}
+		}(node)
+	}
+
+	values, err := utils.Collect(ctx, cancel, errC, valueC)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	var statuses []GravityStatus
+	for _, v := range values {
+		status, ok := v.(GravityStatus)
+		if !ok {
```

Comment:
I know this has been already been discussed and I could just offer my two cents on this by saying I'd be fine with type-cast panicking here since this does not escape the function (technically it does - into a goroutine by that happens in the context of the parent function so is ok) scope and in my book this is a programming error that should be fixed - not logged.
But I think this is a flawed abstraction in the first place - i.e. the use of boxing with `interface{}` only to extract bits into a specific type later. I know there's no way around it with these APIs, but the whole function can be also implemented inline with proper type checking, for example with the use of `errgroup` (but not necessarily):
```go
import "golang.org/x/sync/errgroup"
...
func (c *TestContext) Status(nodes []Gravity) (statuses []GravityStatus, err error) {
	ctx, cancel := context.WithTimeout(c.ctx, c.timeouts.NodeStatus)
	defer cancel()

	valueC := make(chan GravityStatus, len(nodes))
	g, ctx := errgroup.WithContext(ctx)
	for _, node := range nodes {
		node := node
		g.Go(func() error {
			status, err := node.Status(ctx)
			if err != nil {
				return trace.Wrap(err)
			}
			if status != nil {
				valueC <- *status
			}
			return nil
		})
	}
	err = g.Wait()
	if err != nil {
		return nil, trace.Wrap(err)
	}
	close(valueC)
	for status := range valueC {
		statuses = append(statuses, status)
	}
	return statuses, nil
}
```

Reply:
Updated to your suggestion in 0809da1. In addition to strong typing, this code is also way less complex than `utils.Collect` which is a benefit in its own right.

I'll update the testing in a sec once my run finishes.

I'm still interested to hear back if you think `utils.Collect` has value at all (other than offering backend logic for `utils.CollectErrors`).